core/src/main/java/com/linecorp/armeria/internal/common/stream/RecoverableStreamMessage.java

```diff
@@ -141,6 +143,24 @@ public void abort(Throwable cause) {
         }
     }
 
+    @Override
+    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
```

Comment:
It is not related to this PR but overridden to optimize the `aggregate()` API which depends on `collect()`.

Reply:
> It is not related to this PR but overridden to optimize the aggregate() API which depends on collect().

From my understanding, this isn't an optimization but a behavior change.

I'm having a little trouble understanding the purpose of `allowResuming` and why we allow recovery even when `wroteAny` is true here. Would you mind explaining? 😅

Reply:
I don't see `wroteAny` can be `true` in `collect()` since `wroteAny` is only used in `RecoverableSubscriber`.
`allowResuming` means that a `StreamMessage` allows continuing the stream chain after an error.
`super.collect()` uses `StreamMessageCollector` to subscribe to the stream one by one so that the returned future completes all published items in the upstream and recovered items in the fallback together.
The following test is an example of that.
https://github.com/line/armeria/blob/cd08881942d5704da7cdb8b691b62f95488d3dc9/core/src/test/java/com/linecorp/armeria/internal/common/stream/RecoverableStreamMessageTest.java#L73-L78

`upstream.collect()` basically collects all elements and publishes them as a `List`. If an error occurred while writing objects to the stream or the stream is aborted, the stream exceptionally completes the collecting future.
So the downstream can't see any written object at all.

One behavior has changed for `HttpResponse`. Previously, `HttpResponse.aggregate()` subscribes to the `HttpObject`s one by one. Although no object is written to a channel, the stream can not be recovered.
Because `wroteAny` is set to true. If a stream is aggregated at once, the closed stream in the middle of writing can be recovered when using `aggregate().

Reply:
Ah, thanks for the explanation on `wroteAny`.

> allowResuming means that a StreamMessage allows continuing the stream chain after an error.

I see. So `allowResuming` basically means if an error occurred after subscription, the error can be recovered.
In this case, should the condition be inverted since:
1) `allowResuming == true`: `StreamMessage.super.collect` is called which doesn't recover from an exception
2) `allowResuming == false`: The exception is recovered if possible

```
    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
        if (!allowResuming) { <- inverted
            // As the optimized `collect()` method gathers all elements into a list, resuming is unsupported.
            return StreamMessage.super.collect(executor, options);
        }
```

Reply:
> StreamMessage.super.collect is called which doesn't recover from an exception

I think the condition is correct.
`StreamMessage.super.collect` eventually subscribes to the current stream one by one using a `Subscriber`.
So it can relay the items before an error and recover the error with a fallback stream and resume the stream.

>  // As the optimized `collect()` method gathers all elements into a list, resuming is unsupported.

The comment is needed to update to clear the confusion. 😅